do not use collect in describe

bkamins opened the pull request:
Fixes https://github.com/JuliaData/DataFrames.jl/issues/2693

@pdeffebach - some tests (correctness and performance) would be welcome if you can spare some time.

In my opinion this change should be OK, as rather the called functions should add support for `skipmissing` value passed - and I would rather fix this limitation if possible.

nalimilan commented:
> `median` will call `collect` anyway (it does it always).

Yeah actually we could call `median!` if we do that after calling all other functions. That would avoid a copy.

bkamins commented:
if we decided to keep `collect`. In this PR I have removed `collect`.

pdeffebach commented:
I see that there is this warning in `describe`:

> If custom functions are provided, they are called repeatedly with the
  vector corresponding to each column as the only argument. For columns
  allowing for missing values, the vector is wrapped in a call to
  skipmissing: custom functions must therefore support such objects (and
  not only vectors), and cannot access missing values.

To be honest I'm not sure people will read this or follow it. Given there isn't a huge performance improvement, I would err on the side of safety and just use `collect` and get rid of that note.

bkamins commented:
OK - @nalimilan - if you agree to add custom `collect` for `skipmissing` in CategoricalArrays.jl I would just update the docstring in this PR and retain `collect`.

bkamins commented:
I have reverted the commit and just updated the docstring.

nalimilan commented:
Sorry, actually I wonder whether the best solution would be to call `collect` only internally as an optimization when we compute the median (via `median!`), but always pass the `SkipMissing` iterators to custom functions. That way we would still be able to avoid making copies at all when we don't compute the median or quantiles. Otherwise we would be locked into a suboptimal implementation in terms of performance. What do you think?

Here's a simple benchmark:
```julia
julia> df = DataFrame(rand([1:10; missing], 500000, 30), :auto);

# With collect
julia> @btime describe(df, :mean, :min, :max, :nmissing, :eltype);
  210.184 ms (896 allocations: 150.05 MiB)

# Without collect
julia> @btime describe(df, :mean, :min, :max, :nmissing, :eltype);
  76.155 ms (356 allocations: 30.94 KiB)
```

bkamins commented:
Let us do the following, we do `collect` if we:
1. calculate median or quantiles
2. user passes a custom function (which, as @pdeffebach might not support `SkipMissings`)

(i.e. when we use only internal functions that do not need to collect we would not collect)

I will do an update to the PR.

nalimilan commented:
Though that would mean that users cannot benefit from maximum performance if they pass custom functions. Do we really expect that calling `collect` manually would be annoying for users? They should be used to handling `skipmissing` as it's basically needed everywhere.

bkamins commented:
The tension is between:
1. calling `collect` once and then all functions can use `collect`ed value which is faster and collects only once
2. not `calling` collect which means that functions that do not need collect can allocate less memory

Let us wait for @pdeffebach to comment what he things as an end user :smile:.

bkamins commented:
@pdeffebach - could you comment what you feel is preferable here given the discussion we had above? Thank you!

pdeffebach commented:
Oh, sorry I missed this!

I'm fine with not using collect, I guess. I would hate for very large datasets to be super laggy at `describe`. 

I do think that the issue with custom functions is important. People should be able to do other things than just with `skipmissing`. Here's what I propose

1. Don't use `collect` in general. Remove `median` from the default output and only `collect` unless `median` or `:q25` etc. are called. 
2.  When people pass custom functions, pass the whole vector. The user can deal with `missing`s on their own.

pdeffebach commented:
Sorry for bikeshedding with a new proposal. I guess, I don't have a super strong opinion on this. I think reporting the median is a good idea, but understand the appeal of not allocating. 

As for the custom functions, I'm fine with requiring that they work with a `SkipMissing`. You are right, Milan, that they should know how to work with this.

bkamins commented:
OK - so I have just removed `collect` (as `median` will collect it automatically anyway), reverted the docstring (as now it matches what we do) and despecialized some methods to avoid inference time hit of the change we do.

bkamins commented:
Thank you!
